File: MATLABversion/DAS_Package_Kennedy/DAS_beamforming_main.m
% This script is written to use the DAS function
% for beamforming
% with simulated data from kWave

clear all; close all; clc;

% load data
load('simulated_sensor_data_double_pitch_18MHz.mat');
channel_data = sensor_data';


% run DAS algorithm
rf = channel_data;
no_ele = size(rf,2); % number of transducer elements

% Fs = 100*1e6;
channelSpacing = pitch_mm; % [mm] % distance between elements
speedSound = 1540; % [m/s]
sampleSpacing = (1/fs*speedSound)*1000; % for PA data, how far sound travels per time sample
x_axis = [1:size(rf,2)] * channelSpacing; % [mm]
y_axis = [1:size(rf,1)] * sampleSpacing; % [mm]

f_num = 4; % aperture growth may not work well

bf_data = DAS(rf, no_ele, fs, channelSpacing, speedSound);

% visualize the channel data
subplot(1, 2, 1);
imagesc(x_axis, y_axis, channel_data);
xlabel('lateral [mm]');
ylabel('axial [mm]');
axis image;
colormap gray;
colorbar;
title('Before beamforming');

% visualize the data
bf_data_env = abs(hilbert(bf_data));
bf_data_norm = bf_data_env./max(bf_data_env(:));
bf_data_db = db(bf_data_norm);


subplot(1,2,2);
imagesc(x_axis, y_axis, bf_data_db,[-40,0]);
xlabel('lateral [mm]');
ylabel('axial [mm]');
axis image;
colormap gray;
colorbar;
title('After beamforming')

File: DAS_Package_Python/DAS_beamforming_main.py
import numpy as np
import scipy.io as sio
import matplotlib.pyplot as plt
from scipy.signal import hilbert
from DAS import DAS

if __name__ == "__main__":
    data = sio.loadmat('simulated_sensor_data_double_pitch_18MHz.mat')

    sensor_data = data['sensor_data']
    fs = data['fs'].squeeze().item()
    pitch_mm = data['pitch_mm'].squeeze().item()

    channel_data = sensor_data.T
    rf = channel_data

    no_ele = rf.shape[1]  # number of transducer elements

    # Physical parameters
    channel_spacing = pitch_mm  # mm
    speed_sound = 1540  # m/s

    sample_spacing = (1 / fs * speed_sound) * 1000  # mm per sample

    # Axes
    x_axis = np.arange(1, rf.shape[1] + 1) * channel_spacing
    y_axis = np.arange(1, rf.shape[0] + 1) * sample_spacing

    # Run beamforming
    bf_data = DAS(rf, no_ele, fs, channel_spacing, speed_sound)

    # Plot before beamforming
    plt.figure(figsize=(12, 5))

    plt.subplot(1, 2, 1)
    plt.imshow(
        channel_data,
        aspect='auto',
        cmap='gray',
        extent=[x_axis[0], x_axis[-1], y_axis[-1], y_axis[0]]
    )

    plt.xlabel("lateral [mm]")
    plt.ylabel("axial [mm]")
    plt.title("Before beamforming")
    plt.colorbar()

    bf_data_env = np.abs(hilbert(bf_data, axis=0))
    bf_data_norm = bf_data_env / np.max(bf_data_env) # Normalize
    bf_data_db = 20 * np.log10(bf_data_norm + 1e-12) # Convert to dB

    # Plot after beamforming
    plt.subplot(1, 2, 2)
    plt.imshow(
        bf_data_db,
        aspect='auto',
        cmap='gray',
        vmin=-40,
        vmax=0,
        extent=[x_axis[0], x_axis[-1], y_axis[-1], y_axis[0]]
    )

    plt.xlabel("lateral [mm]")
    plt.ylabel("axial [mm]")
    plt.title("After beamforming")
    plt.colorbar()

    plt.tight_layout()
    plt.show()

